commit 498a22333b835a598ccaed4656e97a0ec3573665
Author: Roland McGrath <roland@hack.frob.com>
Date:   Fri Mar 14 10:44:34 2014 -0700

    Compile with -Wundef.

Conflicts:
	Makeconfig
	  (Textual conflict due to glibc-rh1418978-2-6.patch.)
	sysdeps/microblaze/nptl/tls.h
	  (Our tree does not have a microblaze port, removed manually.)

diff --git a/Makeconfig b/Makeconfig
index f23fd93febdf4a86..02f1cc1255c5ac07 100644
--- a/Makeconfig
+++ b/Makeconfig
@@ -643,6 +643,7 @@ ifeq ($(all-warnings),yes)
 else
 +gccwarn := -Wall -Wwrite-strings -Winline
 endif
++gccwarn += -Wundef
 +gccwarn-c = -Wstrict-prototypes -Werror=implicit-function-declaration
 
 # We do not depend on the address of constants in different files to be
diff --git a/include/errno.h b/include/errno.h
index d26a9ee3ff320e3a..7df41dfc3166ef90 100644
--- a/include/errno.h
+++ b/include/errno.h
@@ -6,6 +6,11 @@
 
 # if IS_IN (rtld)
 #  include <dl-sysdep.h>
+#  ifndef RTLD_PRIVATE_ERRNO
+#   error "dl-sysdep.h must define RTLD_PRIVATE_ERRNO!"
+#  endif
+# else
+#  define RTLD_PRIVATE_ERRNO	0
 # endif
 
 # if RTLD_PRIVATE_ERRNO
diff --git a/include/stackinfo.h b/include/stackinfo.h
new file mode 100644
index 0000000000000000..b9b227848753ceb8
--- /dev/null
+++ b/include/stackinfo.h
@@ -0,0 +1,42 @@
+/* Details about the machine's stack: wrapper header.
+   Copyright (C) 2014 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef _INCLUDE_STACKINFO_H
+#define _INCLUDE_STACKINFO_H	1
+
+/* A sysdeps/.../stackinfo.h file defines details for the CPU.
+   It is obliged to define either _STACK_GROWS_DOWN or _STACK_GROWS_UP.  */
+#include_next <stackinfo.h>
+
+#if defined _STACK_GROWS_DOWN && _STACK_GROWS_DOWN
+# ifdef _STACK_GROWS_UP
+#  error "stackinfo.h should not define both!"
+# else
+#  define _STACK_GROWS_UP	0
+# endif
+#elif defined _STACK_GROWS_UP && _STACK_GROWS_UP
+# ifdef _STACK_GROWS_DOWN
+#  error "stackinfo.h should not define both!"
+# else
+#  define _STACK_GROWS_DOWN	0
+# endif
+#else
+# error "stackinfo.h must define _STACK_GROWS_UP or _STACK_GROWS_DOWN!"
+#endif
+
+#endif  /* include/stackinfo.h */
diff --git a/nptl/sysdeps/i386/tls.h b/nptl/sysdeps/i386/tls.h
index 90c7a534bfd0ed16..55877d71d2051d4c 100644
--- a/nptl/sysdeps/i386/tls.h
+++ b/nptl/sysdeps/i386/tls.h
@@ -102,9 +102,6 @@ union user_desc_init
 };
 
 
-/* Get the thread descriptor definition.  */
-# include <nptl/descr.h>
-
 /* This is the size of the initial TCB.  Can't be just sizeof (tcbhead_t),
    because NPTL getpid, __libc_alloca_cutoff etc. need (almost) the whole
    struct pthread even when not linked with -lpthread.  */
@@ -122,6 +119,10 @@ union user_desc_init
 /* The TCB can have any size and the memory following the address the
    thread pointer points to is unspecified.  Allocate the TCB there.  */
 # define TLS_TCB_AT_TP	1
+# define TLS_DTV_AT_TP	0
+
+/* Get the thread descriptor definition.  */
+# include <nptl/descr.h>
 
 
 /* Install the dtv pointer.  The pointer passed is to the element with
diff --git a/nptl/sysdeps/powerpc/tls.h b/nptl/sysdeps/powerpc/tls.h
index 33efc7aea638f138..a04dfa89aa1ec73c 100644
--- a/nptl/sysdeps/powerpc/tls.h
+++ b/nptl/sysdeps/powerpc/tls.h
@@ -49,6 +49,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* We use the multiple_threads field in the pthread struct */
 #define TLS_MULTIPLE_THREADS_IN_TCB	1
@@ -56,6 +57,7 @@ typedef union dtv
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
 
+
 /* The stack_guard is accessed directly by GCC -fstack-protector code,
    so it is a part of public ABI.  The dtv and pointer_guard fields
    are private.  */
diff --git a/nptl/sysdeps/s390/tls.h b/nptl/sysdeps/s390/tls.h
index 4b1ce051c4a87652..580196b7bfee3046 100644
--- a/nptl/sysdeps/s390/tls.h
+++ b/nptl/sysdeps/s390/tls.h
@@ -73,9 +73,6 @@ typedef struct
 /* Get system call information.  */
 # include <sysdep.h>
 
-/* Get the thread descriptor definition.  */
-# include <nptl/descr.h>
-
 /* This is the size of the initial TCB.  Can't be just sizeof (tcbhead_t),
    because NPTL getpid, __libc_alloca_cutoff etc. need (almost) the whole
    struct pthread even when not linked with -lpthread.  */
@@ -93,6 +90,10 @@ typedef struct
 /* The TCB can have any size and the memory following the address the
    thread pointer points to is unspecified.  Allocate the TCB there.  */
 # define TLS_TCB_AT_TP	1
+# define TLS_DTV_AT_TP	0
+
+/* Get the thread descriptor definition.  */
+# include <nptl/descr.h>
 
 
 /* Install the dtv pointer.  The pointer passed is to the element with
diff --git a/nptl/sysdeps/sh/tls.h b/nptl/sysdeps/sh/tls.h
index a7f53dfecf0ca6d6..4659af51c0be7595 100644
--- a/nptl/sysdeps/sh/tls.h
+++ b/nptl/sysdeps/sh/tls.h
@@ -76,6 +76,7 @@ typedef struct
 
 /* The TLS blocks start right after the TCB.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/nptl/sysdeps/sparc/tls.h b/nptl/sysdeps/sparc/tls.h
index cfc9d96c67027caf..2ef0d4e1012259ad 100644
--- a/nptl/sysdeps/sparc/tls.h
+++ b/nptl/sysdeps/sparc/tls.h
@@ -69,9 +69,6 @@ typedef struct
 /* Get system call information.  */
 # include <sysdep.h>
 
-/* Get the thread descriptor definition.  */
-# include <nptl/descr.h>
-
 register struct pthread *__thread_self __asm__("%g7");
 
 /* This is the size of the initial TCB.  Can't be just sizeof (tcbhead_t),
@@ -91,6 +88,10 @@ register struct pthread *__thread_self __asm__("%g7");
 /* The TCB can have any size and the memory following the address the
    thread pointer points to is unspecified.  Allocate the TCB there.  */
 # define TLS_TCB_AT_TP	1
+# define TLS_DTV_AT_TP	0
+
+/* Get the thread descriptor definition.  */
+# include <nptl/descr.h>
 
 /* Install the dtv pointer.  The pointer passed is to the element with
    index -1 which contain the length.  */
diff --git a/nptl/sysdeps/x86_64/tls.h b/nptl/sysdeps/x86_64/tls.h
index cfba4e142ac5dae8..37e9af71b80d18b0 100644
--- a/nptl/sysdeps/x86_64/tls.h
+++ b/nptl/sysdeps/x86_64/tls.h
@@ -91,10 +91,6 @@ typedef struct
 /* Get system call information.  */
 # include <sysdep.h>
 
-
-/* Get the thread descriptor definition.  */
-# include <nptl/descr.h>
-
 #ifndef LOCK_PREFIX
 # ifdef UP
 #  define LOCK_PREFIX	/* nothing */
@@ -120,6 +116,10 @@ typedef struct
 /* The TCB can have any size and the memory following the address the
    thread pointer points to is unspecified.  Allocate the TCB there.  */
 # define TLS_TCB_AT_TP	1
+# define TLS_DTV_AT_TP	0
+
+/* Get the thread descriptor definition.  */
+# include <nptl/descr.h>
 
 
 /* Install the dtv pointer.  The pointer passed is to the element with
diff --git a/ports/sysdeps/hppa/nptl/tls.h b/ports/sysdeps/hppa/nptl/tls.h
index 5eb55a53603bb31e..be14fe6599c9f41b 100644
--- a/ports/sysdeps/hppa/nptl/tls.h
+++ b/ports/sysdeps/hppa/nptl/tls.h
@@ -51,6 +51,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/sysdeps/aarch64/nptl/tls.h b/sysdeps/aarch64/nptl/tls.h
index 05135db57ab69ef6..e08ff732efcc4fa2 100644
--- a/sysdeps/aarch64/nptl/tls.h
+++ b/sysdeps/aarch64/nptl/tls.h
@@ -48,6 +48,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/sysdeps/alpha/nptl/tls.h b/sysdeps/alpha/nptl/tls.h
index 18c29c80fa369e35..f593ba7162954e99 100644
--- a/sysdeps/alpha/nptl/tls.h
+++ b/sysdeps/alpha/nptl/tls.h
@@ -42,6 +42,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/sysdeps/arm/nptl/tls.h b/sysdeps/arm/nptl/tls.h
index 52f695e58b22b406..00b1e53c33064a2e 100644
--- a/sysdeps/arm/nptl/tls.h
+++ b/sysdeps/arm/nptl/tls.h
@@ -49,6 +49,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/sysdeps/ia64/nptl/tls.h b/sysdeps/ia64/nptl/tls.h
index 77574e55db890043..f270f958bcfc53de 100644
--- a/sysdeps/ia64/nptl/tls.h
+++ b/sysdeps/ia64/nptl/tls.h
@@ -87,6 +87,7 @@ register struct pthread *__thread_self __asm__("r13");
 
 /* The DTV is allocated at the TP; the TCB is placed elsewhere.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/sysdeps/m68k/nptl/tls.h b/sysdeps/m68k/nptl/tls.h
index 8b0909237b303695..7f4bbfb9162dd610 100644
--- a/sysdeps/m68k/nptl/tls.h
+++ b/sysdeps/m68k/nptl/tls.h
@@ -49,6 +49,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/sysdeps/mach/hurd/i386/tls.h b/sysdeps/mach/hurd/i386/tls.h
index d0a72f0e5c032820..8619820d3b0502b5 100644
--- a/sysdeps/mach/hurd/i386/tls.h
+++ b/sysdeps/mach/hurd/i386/tls.h
@@ -26,6 +26,7 @@
 /* The TCB can have any size and the memory following the address the
    thread pointer points to is unspecified.  Allocate the TCB there.  */
 #define TLS_TCB_AT_TP	1
+#define TLS_DTV_AT_TP	0
 
 #ifndef __ASSEMBLER__
 
diff --git a/sysdeps/mips/nptl/tls.h b/sysdeps/mips/nptl/tls.h
index 8a97437bad4752c4..669a79c0c64cb566 100644
--- a/sysdeps/mips/nptl/tls.h
+++ b/sysdeps/mips/nptl/tls.h
@@ -62,6 +62,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* Get the thread descriptor definition.  */
 # include <nptl/descr.h>
diff --git a/sysdeps/tile/nptl/tls.h b/sysdeps/tile/nptl/tls.h
index 303a95e000fee30f..132de9021e5651f0 100644
--- a/sysdeps/tile/nptl/tls.h
+++ b/sysdeps/tile/nptl/tls.h
@@ -49,6 +49,7 @@ typedef union dtv
 
 /* The TP points to the start of the thread blocks.  */
 # define TLS_DTV_AT_TP	1
+# define TLS_TCB_AT_TP	0
 
 /* We use the multiple_threads field in the pthread struct */
 #define TLS_MULTIPLE_THREADS_IN_TCB	1
